### PR TITLE
Pager does not generate many duplicate where conditions and order by …

### PR DESCRIPTION
**Pull request by mightyguava**
…columns

Adds a `clone()` method to the Query so that the Pager can modify a copy
of the query instead of the original.

Fixes https://github.com/cashapp/misk/issues/1613

### CODE REVIEW

File: misk-hibernate/src/test/kotlin/misk/hibernate/pagination/RealPagerTest.kt [hunk 1]
```diff
@@ -10,19 +10,23 @@ import misk.hibernate.Query
...
+import java.time.Instant
 import java.time.LocalDate
 import javax.inject.Inject
 
 @MiskTest(startService = true)
 class RealPagerTest {
 
   @MiskTestModule
-  val module = MoviesTestModule()
+  val module = MoviesTestModule(DataSourceType.MYSQL)
```
Comment (mightyguava) [inline misk-hibernate/src/test/kotlin/misk/hibernate/pagination/RealPagerTest.kt:29]: The default `VITESS_MYSQL` data source doesn't like `SET GLOBAL`, used to enable the general query log

File: misk-hibernate/src/main/kotlin/misk/hibernate/ReflectionQueryFactory.kt [hunk 2]
```diff
@@ -68,6 +68,28 @@ internal class ReflectionQuery<T : DbEntity<T>>(
     disabledChecks += check
   }
 
+  override fun <Q: Query<*>> clone(): Q {
```
Comment (mightyguava) [inline misk-hibernate/src/main/kotlin/misk/hibernate/ReflectionQueryFactory.kt:71]: This is a scary hack but it works and I couldn't think of a better way of doing thiis

Comment (zcross) [reply misk-hibernate/src/main/kotlin/misk/hibernate/ReflectionQueryFactory.kt:71]: Mind adding a simple regression test?

File: misk-hibernate/src/test/kotlin/misk/hibernate/pagination/RealPagerTest.kt [hunk 3]
```diff
@@ -120,6 +124,62 @@ class RealPagerTest {
...
+      }
+    }
+
+    val pageQueries = queries.reversed().filter { it.contains("select") }.map {
+      Regex("""/\* (.*) \*/""").find(it)!!.groups[1]!!.value
+    }
+    assertThat(pageQueries).hasSize(3)
+    // The first page doesn't have an (id > ?) condition
+    assertThat(countClauses(pageQueries[0])).isEqualTo(Clauses(1, 1))
+    // The second page each have exactly one (id > ?) condition and one ORDER BY column
```
Comment (zcross) [inline misk-hibernate/src/test/kotlin/misk/hibernate/pagination/RealPagerTest.kt:163]: Excellent test coverage

File: misk-hibernate/src/main/kotlin/misk/hibernate/pagination/RealPager.kt [hunk 4]
```diff
@@ -19,6 +20,8 @@ internal class RealPager<T : DbEntity<T>, Q: Query<T>>(
   }
 
   override fun nextPage(session: Session): Page<T>? {
+    @Suppress("UNCHECKED_CAST")
+    val query = query.clone<Q>()
```
Comment (zcross) [inline misk-hibernate/src/main/kotlin/misk/hibernate/pagination/RealPager.kt:24]: WDYT about renaming either the local val or the "original" query val?

**Review by zcross — APPROVED** [review]
LGTM, thanks for fixing so this soon after spotting the symptoms

File: misk-hibernate/src/main/kotlin/misk/hibernate/pagination/RealPager.kt (hunk 4 above)
Comment (mightyguava) [reply misk-hibernate/src/main/kotlin/misk/hibernate/pagination/RealPager.kt:24]: I did this to purposefully shadow the "original" query val, so you can't accidentally touch it in this function

File: misk-hibernate/src/main/kotlin/misk/hibernate/ReflectionQueryFactory.kt (hunk 2 above)
Comment (mightyguava) [reply misk-hibernate/src/main/kotlin/misk/hibernate/ReflectionQueryFactory.kt:71]: Yeah good idea, added a test

**Review by swankjesse — COMMENTED** [review]
Nice fix. Thank you

**Pull request merged**